src/tree-view/keyboard-navigation/index.tsx

```diff
@@ -0,0 +1,302 @@
+    setTimeout(() => {
+      if (this._treeView) {
+        // Update focused tree-item in case tree-items change.
+        this.updateFocusedTreeItem(this.focusedTreeItem?.element);
+        this._navigationAPI.current?.updateFocusTarget();
+      }
+    }, 0);
+  }
+
+  public update({ pageUpDownSize }: { pageUpDownSize: number }) {
+    this._pageUpDownSize = pageUpDownSize;
+  }
+
+  public getNextFocusTarget = () => {
+    if (!this.treeView) {
+      return null;
+    }
+
+    const treeItem = this.focusedTreeItem;
+    const firstTreeItemToggle = this.treeView.querySelector(
+      '[data-awsui-tree-view-toggle-button=true]'
+    ) as null | HTMLButtonElement;
+
+    let focusTarget: null | HTMLElement = firstTreeItemToggle;
+
+    // Focus on the element that was focused before.
+    if (treeItem) {
+      focusTarget = this.getNextFocusableTreeItem(treeItem, 0);
+    }
+
+    return focusTarget;
+  };
+
+  private get pageUpDownSize() {
+    return this._pageUpDownSize;
+  }
+
+  private get treeView(): null | HTMLUListElement {
+    return this._treeView;
+  }
+
+  private getFocusablesFrom(target: HTMLElement) {
+    const isElementRegistered = (element: Element) => this._navigationAPI.current?.isRegistered(element);
+    return getAllFocusables(target).filter(el => isElementRegistered(el) && !isElementDisabled(el));
+  }
+
+  private isRegistered(element: null | Element): boolean {
+    return !element || (this._navigationAPI.current?.isRegistered(element) ?? false);
+  }
+
+  private updateFocusedTreeItem(focusedElement?: HTMLElement): void {
+    if (!focusedElement) {
+      return;
+    }
+
+    const treeItem = getClosestTreeItem(focusedElement);
+    if (!treeItem) {
+      return;
+    }
+    const treeItemIndex = parseInt(treeItem.getAttribute('data-awsui-tree-item-index') ?? '');
+
+    this.focusedTreeItem = {
+      treeItemId: treeItem.id,
+      treeItemIndex,
+      element: focusedElement,
+    };
+  }
+
+  private onFocusin = (event: FocusEvent) => {
+    if (!(event.target instanceof HTMLElement)) {
+      return;
+    }
+
+    this.updateFocusedTreeItem(event.target);
+    if (!this.focusedTreeItem) {
+      return;
+    }
+
+    this._navigationAPI.current?.updateFocusTarget();
+  };
+
+  private setNavigationActive(navigationActive: boolean) {
+    this.set(() => ({ navigationActive }));
+  }
+
+  private onFocusout = (event: FocusEvent) => {
+    const target = event.relatedTarget as null | Element;
+
+    // Re-enable navigation
+    // so that when tree-view is focused again, the last focused tree-item can be re-focused
+    if (target === null) {
+      this.setNavigationActive(true);
+      return;
+    }
+
+    // Disable navigation if the new focus target is inside the tree-item content
+    // so that the focusables inside can be navigated with tab
+    const closestTreeItemContent = getClosestTreeItemContent(target);
+    if (closestTreeItemContent) {
```

Comment:
nit: this can be simplified to `this.setNavigationActive(!closestTreeItemContent)`